### ISSUE TITLE
C++ binary compiled with -m32 doesn't run outside build environment

### ISSUE THREAD

**Issue by musteresel** [NixOS/nixpkgs#70179]
**Describe the bug**
C++ code compiled on x86-64 with -m32 flag (in multiStdenv) runs in the build environment but not when installed. C code (with -m32 flag) on the other hand does run in both environments.

**To Reproduce**
1. Create `mwe.nix`

~~~
with import <nixpkgs> {};

multiStdenv.mkDerivation {
  name = "mwe";
  unpackPhase = "true";
  buildPhase =
    ''
      mkdir logs
      echo "int main() {}" > tst.cc
      $CXX -m32 -o tst-cxx tst.cc &> logs/cxx.txt
      $CC -m32 -o tst-cc tst.cc &> logs/cc.txt
    '';
  checkPhase =
    ''(
      set -ex
      ./tst-cc
      ./tst-cxx
      )
    '';
  doCheck = true;
  installPhase =
    ''
      mkdir -p $out/
      cp tst-* $out/
      cp -r logs $out/
    '';
}
~~~

2. Build with `nix-build mwe.nix` (same issue when adding `-I nixpkgs=https://github.com/NixOS/nixpkgs/archive/19.09-beta.tar.gz`). Output:

    ~~~
   these derivations will be built:
     /nix/store/1b7pfm9znmrnjcyk78d8083kghxpxlql-mwe.drv
   building '/nix/store/1b7pfm9znmrnjcyk78d8083kghxpxlql-mwe.drv'...
   unpacking sources
   patching sources
   configuring
   no configure script, doing nothing
   building
   running tests
   ++ ./tst-cc
   ++ ./tst-cxx
   installing
   post-installation fixup
   shrinking RPATHs of ELF executables and libraries in /nix/store/8cjp4a90w6mxb2f0wx44cwk0rfkah3rw-mwe
   shrinking /nix/store/8cjp4a90w6mxb2f0wx44cwk0rfkah3rw-mwe/tst-cc
   shrinking /nix/store/8cjp4a90w6mxb2f0wx44cwk0rfkah3rw-mwe/tst-cxx
   strip is /nix/store/0y7jmqnj48ikjh37n3dl9kqw9hnn68nq-binutils-2.31.1/bin/strip
   patching script interpreter paths in /nix/store/8cjp4a90w6mxb2f0wx44cwk0rfkah3rw-mwe
   checking for references to /build/ in /nix/store/8cjp4a90w6mxb2f0wx44cwk0rfkah3rw-mwe...
   /nix/store/8cjp4a90w6mxb2f0wx44cwk0rfkah3rw-mwe
    ~~~~

3. Run `(set -x; ./result/tst-cc; ./result/tst-cxx )`. Output:

    ~~~
    + ./result/tst-cc
    + ./result/tst-cxx
    ./result/tst-cxx: error while loading shared libraries: libstdc++.so.6: wrong ELF class: ELFCLASS64
    ~~~

**Expected behavior**
Successful build and run of the two executables.

**Additional context**
This is a Minimal Working Example, the issue arises for me with a large codebase which must be compiled with `-m32`

**Metadata**
Result of `nix run nixpkgs.nix-info -c nix-info -m`:
 - system: `"x86_64-linux"`
 - host os: `Linux 5.2.14, NixOS, 19.03.173481.8a30e242181 (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2.2`
 - channels(musteresel): `""`
 - channels(root): `"nixos-19.03.173481.8a30e242181"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by musteresel** [comment]
[result/logs/cc.txt](https://github.com/NixOS/nixpkgs/files/3676772/cc.txt)
[result/logs/cxx.txt](https://github.com/NixOS/nixpkgs/files/3676773/cxx.txt)

**Comment by musteresel** [comment]
Adding `dontPatchELF = true;` to the `mwe.nix` "fixes" this issue; thus this is seems to be a problem with patchelf.

Manually running `patchelf --debug --shrink-rpath tst-cxx` with both the `${pkgs.patchelf}` and `${pkgsi686Linux.patchelf}` shows the same result; it shrinks the rpath and removes the required path for libstdc++.so.6:

~~~
# executed during nix-build buildPhase
$ ldd tst-cxx
        linux-gate.so.1 (0xf7fd4000)
        libstdc++.so.6 => /nix/store/yfcj76xvkmzxd5jri3l76n4a5gc93xwp-gcc-7.4.0-lib/lib/libstdc++.so.6 (0xf7e49000)
        libm.so.6 => /nix/store/v1dks37jzwmy7r7sfd1famp99508sln6-glibc-multi-2.27/lib/32/libm.so.6 (0xf7d46000)
        libgcc_s.so.1 => /nix/store/yfcj76xvkmzxd5jri3l76n4a5gc93xwp-gcc-7.4.0-lib/lib/libgcc_s.so.1 (0xf7d28000)
        libc.so.6 => /nix/store/v1dks37jzwmy7r7sfd1famp99508sln6-glibc-multi-2.27/lib/32/libc.so.6 (0xf7b54000)
        /nix/store/v1dks37jzwmy7r7sfd1famp99508sln6-glibc-multi-2.27/lib/32/ld-linux.so.2 (0xf7fd5000)

$ patchelf --debug --shrink-rpath tst-cxx
patching ELF file `tst-cxx'
Kernel page size is 4096 bytes
removing directory `/nix/store/x11ar1k58y1m5zkddwa48mpd788p0jzb-mwe2/lib64' from RPATH
removing directory `/nix/store/x11ar1k58y1m5zkddwa48mpd788p0jzb-mwe2/lib' from RPATH
removing directory `/nix/store/yfcj76xvkmzxd5jri3l76n4a5gc93xwp-gcc-7.4.0-lib/lib' from RPATH
removing directory `/nix/store/v1dks37jzwmy7r7sfd1famp99508sln6-glibc-multi-2.27/lib/32' from RPATH
new rpath is `/nix/store/v1dks37jzwmy7r7sfd1famp99508sln6-glibc-multi-2.27/lib:/nix/store/yfcj76xvkmzxd5jri3l76n4a5gc93xwp-gcc-7.4.0-lib/lib64'

$ ldd tst-cxx
        linux-gate.so.1 (0xf7fd4000)
        libstdc++.so.6 => not found
        libm.so.6 => /nix/store/sjsvwd9sqxjflmslnf8xs48aqpkb2sjy-glibc-2.27/lib/libm.so.6 (0xf7ecc000)
        libgcc_s.so.1 => /nix/store/sjsvwd9sqxjflmslnf8xs48aqpkb2sjy-glibc-2.27/lib/libgcc_s.so.1 (0xf7eaf000)
        libc.so.6 => /nix/store/sjsvwd9sqxjflmslnf8xs48aqpkb2sjy-glibc-2.27/lib/libc.so.6 (0xf7cdb000)
        /nix/store/v1dks37jzwmy7r7sfd1famp99508sln6-glibc-multi-2.27/lib/32/ld-linux.so.2 (0xf7fd5000)

~~~

**Comment by musteresel** [comment]
I'm using `patchelf 0.9`, this is fixed in at least `patchelf 0.10-pre-20180509` ... so this should be fixed as soon as `patchelf 0.10` hits the repo (I guess that's a massive rebuild, though, so probably only for the next release?)

**Comment by musteresel** [comment]
Leaving a note if someone stumbles over this and tries to create a 32 bit C++ binary:

`pkgsi686Linux.stdenv` doesn't even add 64 bit library directories to the rpath, thus when using this instead of `stdenvMulti` the issue with patchelf is circumvented
